src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/PipelineAttacher.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2017 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.junitrealtimetestreporter;
+
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.model.Run;
+import hudson.tasks.junit.JUnitParser;
+import hudson.tasks.junit.JUnitResultArchiver;
+import hudson.tasks.test.TestResult;
+import java.io.IOException;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.structs.describable.DescribableModel;
+import org.jenkinsci.plugins.structs.describable.UninstantiatedDescribable;
+import org.jenkinsci.plugins.workflow.FilePathUtils;
+import org.jenkinsci.plugins.workflow.actions.ArgumentsAction;
+import org.jenkinsci.plugins.workflow.actions.WorkspaceAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.flow.GraphListener;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.StepNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.DepthFirstScanner;
+
+/**
+ * Attaches distinct test result actions to each {@code node} (or {@code ws}) block in a job with this plugin activated.
+ * Looks for all {@code junit} steps in the {@code lastSuccessfulBuild} so it can guess at test patterns to archive.
```

Comment:
I am likely missing something obvious but why not looking at the active configuration?

Reply:
There is no “active configuration” for a Pipeline job.

```groovy
library 'open-research-problems'
node {
  sh 'mvn test'
  junit findCounterexampleToGoldbachConjecture() ? '**/TEST-*a.xml' : '**/TEST-*b.xml'
}
```

Reply:
Right, I guess we have to live with the chance the results will reflect outdated config. Which is not a big deal if the implementation have the same side effect as for traditional build  types of re-parsing the results by the actual `junit` steps so the results can be wrong only during the run but not after completion.

---

I realize I am late for the party but perhaps it would be easier to implement this as separate block step:

```
junitRealtime(include: "**/*.xml", ...) {
  // Whatever actions generating report files.
}
```

This way we would know the exact configuration while doing the parsing and we do not have to guess when to start parsing as the generating process is denoted explicitly.

Reply:
> if the implementation have the same side effect as for traditional build types of re-parsing the results by the actual `junit` steps so the results can be wrong only during the run but not after completion

That is the case.

> perhaps it would be easier to implement this as separate block step

That would easier in some ways, yes. My main reservation is that it would duplicate some of the configuration of `JUnitResultArchiver`, though this could be alleviated by having it actually inline that:

```groovy
junitRealtime(junit '**/TEST-*.xml') {
  sh 'mvn test'
}
```

where the end of the block would delete the temporary results and delegate to `JUnitResultArchiver.perform` for the official version.

If you want go that route, I would throw out most of the work in this PR and start over. But I would like to get feedback from @abayer on how a feature like this would mesh with what he is doing.

Reply:
I hate to block this as it is a good change but I hope it can be made more straightforward.

@jenkinsci/code-reviewers, thoughts?